pgjdbc/src/main/java/org/postgresql/core/v3/ConnectionFactoryImpl.java

```diff
@@ -725,6 +727,66 @@ private static String getAuthenticationMethodName(int authReq) {
     }
   }
 
+  private enum AuthMethod {
+    NONE, PASSWORD, MD5, GSS, SSPI, SCRAM_SHA_256;
+
+    static AuthMethod fromString(String method) throws PSQLException {
+      switch (method) {
+        case "none": return NONE;
+        case "password": return PASSWORD;
+        case "md5": return MD5;
+        case "gss": return GSS;
+        case "sspi": return SSPI;
+        case "scram-sha-256": return SCRAM_SHA_256;
+        default: throw new PSQLException(GT.tr("Invalid authentication method: {0}", method), PSQLState.INVALID_PARAMETER_VALUE);
+      }
+    }
+
+    static EnumSet<AuthMethod> parseRequireAuth(String requireAuth) throws PSQLException {
+      if (requireAuth == null) {
+        return EnumSet.noneOf(AuthMethod.class);
+      }
+
+      EnumSet<AuthMethod> allowedMethods;
+      EnumSet<AuthMethod> seenMethods = EnumSet.noneOf(AuthMethod.class);
+      String[] methods = requireAuth.split(",");
+      boolean isDisallowMode = methods.length > 0 && methods[0].trim().startsWith("!");
+
+      allowedMethods = isDisallowMode ? EnumSet.allOf(AuthMethod.class) : EnumSet.noneOf(AuthMethod.class);
+
+      for (String method : methods) {
+        method = method.trim();
+        boolean isNegative = method.startsWith("!");
+
+        if (isNegative != isDisallowMode) {
+          throw new PSQLException(GT.tr("requireAuth cannot mix positive and negative authentication methods"), PSQLState.INVALID_PARAMETER_VALUE);
+        }
+
+        AuthMethod authMethod = fromString(isNegative ? method.substring(1) : method);
+        if (!seenMethods.add(authMethod)) {
+          throw new PSQLException(GT.tr("requireAuth contains duplicate authentication method"), PSQLState.INVALID_PARAMETER_VALUE);
+        }
+
+        if (isDisallowMode) {
+          allowedMethods.remove(authMethod);
+        } else {
+          allowedMethods.add(authMethod);
+        }
+      }
+      return allowedMethods;
+    }
+
+    static void isAllowed(@Nullable String requireAuth, AuthMethod authMethod) throws PSQLException {
```

Comment:
```suggestion
    static void checkAuth(Set<AuthMethod> allowedMethods, AuthMethod authMethod) throws PSQLException {
```

`isAllowed(...)` results in repeated parsing, so why don't we parse the set once and pass it over the methods?

I'm not fond of `is...` methods that return void.

Reply:
`AUTH_REQ_OK` brings the second parse, doesn't it?

Frankly, I find it much eaiser to reason about if the inputs are parsed once, and the code deals with parsed values rather than parse at multiple locations.